### ISSUE TITLE
nx.lowest_common_ancestor incorrectly returns None for this DAG

### ISSUE THREAD

**Issue by odel4y** [networkx/networkx#4942]
### Current Behavior
When executing the `nx.lowest_common_ancestor` algorithm for the pair (**B**, **D**) in this graph, NetworkX incorrectly returns None.

![ancestor-graph](https://user-images.githubusercontent.com/10086888/123952621-88ed8c80-d9a6-11eb-94e7-dafdd4badb10.png)

### Expected Behavior
If I'm not wrong, the lowest common ancestor should exist and be **C**.

### Steps to Reproduce
```python3
import networkx as nx

G = nx.DiGraph()

G.add_edge("A", "B")
G.add_edge("C", "A")
G.add_edge("C", "D")
G.add_edge("E", "A")
G.add_edge("F", "C")

# Should return "C", but returns None
print(nx.lowest_common_ancestor(G, "B", "D", default=None))
```

### Environment

Python version: 3.8.5
NetworkX version: 2.6rc2 (also tested with earlier versions)

**Comment by rossbar** [comment]
I confirm that I can reproduce this on `main` (7fe275fd)

**Comment by dschult** [comment]
Very strange.  After running the above to verify the bug, I then remove the node "E" and it works and then add the edge ("E","A") and it still works.

```python
import networkx as nx

G = nx.DiGraph()

G.add_edge("A", "B")
G.add_edge("C", "A")
G.add_edge("C", "D")
G.add_edge("E", "A")
G.add_edge("F", "C")

# Should return "C", but returns None
nx.lowest_common_ancestor(G, "B", "D", default=None)

G.remove_node("E")
nx.lowest_common_ancestor(G, "B", "D", default=None)  # output 'C'

G.add_edge("E", "A")
nx.lowest_common_ancestor(G, "B", "D", default=None)  # output 'C'
```

So, I suspect node order is playing a role here.  removing and adding 'E' changes the order of nodes 'F' and 'E', and changes the edge order as well.  hmmm...

**Comment by odel4y** [comment]
**TL;DR**: I have so far not found a bug in the code compared to the paper's algorithm. Just following the algorithm from the paper, I already encounter this problem. I hope I overlooked something ...

After some experimentation and stepping through the code, I believe that this bug always appears iff a parent node of one of the two query nodes has two parents itself. The minimal example would be to query the LCA of either (A,D) or (B,D) from this graph (DAG is directed from top to bottom):
```
A  B
\ /
 C
 |
 D
```
I studied the section of the [paper](https://www3.cs.stonybrook.edu/~bender/pub/JALG05-daglca.pdf) the code is based on (_4.3. A Simpler Ancestor-List-Based All-Pairs LCA algorithm_). The general idea is:

1. If the DAG _G_ has multiple source nodes, create an artificial root node and connect it to each source.
2. Generate a spanning tree _T_ over _G_.
3. From the edges present in _G_ but not in _T_ and from all nodes in _G_, create another DAG _D_ .
3. For each query node, create a list containing itself and its ancestors in _D_.
4. For each pair in the two query node lists, compute their Tree LCA in the spanning tree T.
5. The DAG's LCA is the Tree LCA with maximum depth in T.

Let's use the example above and query the nodes (B,D):
```
Query nodes: B, D

Step      1.      2.      3.        4.

Graph/    G       T       D     
Tree         
------------------------------------------

         root    root     B     L_B = {B}
         / \     / \      |
         A B     A B      C
         \ /     \
          C       C       A     
          |       |
          D       D       D     L_D = {D}
```
For the spanning tree _T_, there exist two possible variants: The first one is shown above. We can see that in this spanning tree, there is no possible combination of query nodes (q1 != q2), where the Tree LCA would result in B. The algorithm thus does not work in this case.

The second possible variant of _T_ is:
```
root
/ \
A B
  /
 C
 |
 D
```
In this case the tree LCA of (B,D) correctly returns B. The spanning tree in the code is generated using a DFS traversal of _G_. Which variant we get, is dependent on the node order and thus pure luck.

**Comment by odel4y** [comment]
I believe, the algorithm in NetworkX was originally implemented by @calmofthestorm. Even if it was quite a while ago, maybe he can offer his opinion on the matter :)

**Comment by calmofthestorm** [comment]
I remember not really understanding why you could just pick any spanning tree when I was implementing this, but it worked for every input I fed it, including some rather large random tests I never committed to networkx, and not having a background in graph theory, I was content to trust the authors combined with randomized testing. And to be honest the project I needed this for was...tolerant to the occasional error shall we say -- I wouldn't have noticed any coming up in actual use so long as it was mostly accurate most of the time on nltk ontologies.

It did not occur to me to do exhaustive random tests on simple graphs -- instead I generated large graphs and only ever used the DFS spanning tree. If we ever do find a fix it would be good to add a test that generates all spanning trees over a simple graph.

This is all based on memory; I haven't reread the paper since, and haven't re-read the code since it was finally submitted in 2017.

IIRC most of Bender et al is devoted to describing an asymptotically more efficient algorithm than the one implemented here. The O(n^3) algorithm I implemented is included as well, described relatively briefly, as a suggested to implementers.

So if there is an error, it's possible that the core algorithm is correct, and only the simplified version is wrong.

It's tempting to try to duct tape on a fix to paper over this particular case, but without an understanding of why the algorithm works or why it fails in this case, I think that would be somewhat irresponsible. Pretty sure considering all spanning trees would work, but that's going to get expensive.

It may be worth contacting the authors, if the bug is in the algorithm itself rather than our code. We may be interpreting it wrong, or there may be a simple fix. I would need to reread the paper carefully before I'd feel comfortable doing so personally, which unfortunately I don't think I can commit to these days, but it sounds like you have already done so based on the excellent write up above, so it may be worth a try.

Of course it's been 18 years for them, vs just 8 for me, and I still remember little:-)

**Comment by calmofthestorm** [comment]
Thinking about this some more, is the problem the added edges for the root node? Would it suffice to consider one arbitrary spanning tree per source?

**Comment by odel4y** [comment]
Thanks for your response! Intuitively your approach of creating one spanning tree per source makes sense. This guarantees that the two query nodes are always contained as a pair in at least one of the trees and are connected by an actual path. This way the Tree LCA should always find the respective DAG LCA in at least one of the trees. Empirically it solves the problem for this graph and others I could come up with manually. However, it increases the complexity and probably makes the ancestor lists obsolete.

My search of other implementations of the O(n^3) algorithm only yielded two results which implement the slower [Naive LCA](https://jgrapht.org/javadoc-1.4.0/org/jgrapht/alg/lca/NaiveLCAFinder.html) and the faster [All-Pairs LCA Algorithm](https://github.com/jeremy-murphy/graph-ancestry). So I could not compare the results...

I also thought about contacting the authors. But I will definitely make sure to have thought it through before doing so :)

**Comment by calmofthestorm** [comment]
I really wish they would have explained the ordered ancestor lists thing better.

Does the all-pairs LCA you link to work on graphs? the readme says tree only. I believe it's the same algorithm used as a subroutine in my implementation.

**Comment by dschult** [comment]
In the Bender paper, there must be a typo or mistake in the statement of either the definition of L_x or the definition of the DAG D.  I think L_x should be the ancestors of x in G.  I believe that still satisfies Lemma 4.3!!
Then the same pruning of the all-pairs LCA_T computation brings that from O(n^4) down to O(n^3). So, the complexity is the same even though the number of ancestors stored in the lists is larger.

What do you think?   Does changing line 276 to `           my_ancestors = nx.dag.ancestors(dag, v)` make the code work?

I'm fine with moving to another paper since this one clearly has some troubles. But it has some nice numerical experiments that show it works... so I think there must be a typo of some sort.  I tried looking for the cited code implementing this "simpler" algorithm. If you want to contact an author, I would suggest that person who hosted the code (Pavel Sumazin) who is now at Baylor College of Medicine.  Contactig Bender at SUNY Stony Brook is also good of course.

**Comment by odel4y** [comment]
@calmofthestorm I am not quite sure. The [code](https://github.com/jeremy-murphy/graph-ancestry/blob/master/lowest_common_ancestor.hpp) contains a description that includes DAGs. But I cannot find the actual implementation of the function `lowest_common_ancestor_query`.

@dschult: You are right that this change solves the issue (for all my example trees)! In this case, the additional DAG _D_ would be unnecessary. I am no expert about the computational complexity but would expect that the pruning still works to the same effect.

Indeed it would be interesting to look at the implementation of the authors. I will try to contact them.

**Comment by jtrim-ons** [comment]
I'm in favour of using an algorithm from the literature - either a corrected version of the Bender algorithm or an algorithm from another paper. But I was thinking that it might be a good idea to have a naive implementation too, for testing the main algorithm and perhaps also as a stop-gap until we have a correct, fast implementation.

This is the implementation I came up with (with `all_pairs_lowest_common_ancestor()` being the entry point). I based this on the brute-force algorithm [described in Wikipedia](https://en.wikipedia.org/wiki/Lowest_common_ancestor#Extension_to_directed_acyclic_graphs).

```python
import random
import time
import itertools
import networkx as nx

def find_successor(G, v, S):
    """Find a successor of v that is in the set of nodes S"""
    for w in G.successors(v):
        if w in S:
            return w
    return None

def get_a_lowest_common_ancestor(G, common_ancestors):
    """Return a lowest common ancestor from the given set of common ancestors"""
    # Start with an arbitrary node v from the set of common ancestors.  Follow
    # arbitrary outgoing edges, remaining in the set of common ancestors, until
    # reaching a node with no outgoing edge to another of the common ancestors.
    v = next(iter(common_ancestors))
    while True:
        successor = find_successor(G, v, common_ancestors)
        if successor is None:
            return v
        v = successor

def get_ancestor_set(G, v):
    """Return a set containing v and all of its ancestors in G"""
    ancestor_set = set([v])
    for _, w in nx.bfs_edges(G, v, reverse=True):
        ancestor_set.add(w)
    return ancestor_set

def all_pairs_lowest_common_ancestor(G, pairs=None):
    """Return the lowest common ancestor of all pairs or the provided pairs"""
    result = []
    if pairs is None:
        for i, v in enumerate(G.nodes()):
            v_ancestor_set = get_ancestor_set(G, v)
            for j, w in enumerate(G.nodes()):
                if j > i:
                    break
                w_ancestor_set = get_ancestor_set(G, w)
                common_ancestors = v_ancestor_set & w_ancestor_set
                if common_ancestors:
                    result.append(((v, w), get_a_lowest_common_ancestor(G, common_ancestors)))
    else:
        for v, w in pairs:
            v_ancestor_set = get_ancestor_set(G, v)
            w_ancestor_set = get_ancestor_set(G, w)
            common_ancestors = v_ancestor_set & w_ancestor_set
            if common_ancestors:
                result.append(((v, w), get_a_lowest_common_ancestor(G, common_ancestors)))
    return result


# DAG generator based on https://stackoverflow.com/a/13546785/3347737
n = 200
p = 0.2
random.seed(1)
G=nx.gnp_random_graph(n,p,directed=True)
G = nx.DiGraph([(u,v) for (u,v) in G.edges() if u<v])

start_time = time.time()
lca_count = len([item for item in nx.all_pairs_lowest_common_ancestor(G)])
end_time = time.time()
print("Current algorithm: {:.3f} seconds, {} LCAs".format(end_time - start_time, lca_count))

start_time = time.time()
lca_count = len([item for item in all_pairs_lowest_common_ancestor(G)])
end_time = time.time()
print("Naive algorithm: {:.3f} seconds, {} LCAs".format(end_time - start_time, lca_count))
```

Here are results from one run:
```
Current algorithm: 2.462 seconds, 20011 LCAs
Naive algorithm: 3.409 seconds, 20040 LCAs
```

So, the algorithm I've pasted above is just a little slower and finds some extra lowest common ancestors, at least on this one graph.

Let me know if you'd like me to put this in a pull request in some form.

But to reiterate, I do think a published algorithm would be better. (And I'd love to know what the correct version of the Bender algorithm is.)

**Comment by jamestrimble** [comment]
An update on the algorithm in the previous comment (from my other GitHub account - sorry for being confusing!):

If you replace the function `get_ancestor_set` in the big code listing above with the following version, the naive algorithm's speed seems to be competitive with the existing algorithm.

```python
def get_ancestor_set(G, v):
    """Return a set containing v and all of its ancestors in G"""
    ancestor_set = set([v])
    to_visit = [v]
    while to_visit:
        w = to_visit.pop()
        for u in G.predecessors(w):
            if u not in ancestor_set:
                ancestor_set.add(u)
                to_visit.append(u)
    return ancestor_set
```

**Comment by jamestrimble** [comment]
One last thought on this: the naive algorithm could be further sped up by caching the ancestor list of each vertex. Perhaps this caching feature could be made optional so the user can choose a slower method that requires less memory.

**Comment by dschult** [comment]
Any idea how much the savings in time or memory is? Seems like it would be potentially storing a set of all nodes for each node which would be O(n^2) in memory -- which is on the order of a dense version of the DiGraph itself. But what is the time savings?  Could be substantial I think because you have to do the lookup for each node of each pair -- so you are collecting the ancestors n times.  Since this is a naive algorithm, it seems like collecting a dict of ancestor sets at the beginning might be a good way to go.

**Comment by jamestrimble** [comment]
@dschult I've now implemented a memoised version of the `get_ancestor_set()` function. Here are the timings for one run with n=300 and p=0.1:

```
Current algorithm: 11.491 seconds, 44481 LCAs
Naive algorithm with cache: 0.465 seconds, 44674 LCAs
Naive algorithm: 5.465 seconds, 44674 LCAs
```

So the cache gives an order of magnitude improvement in speed, at least for this sample of size 1 :-)

As you say, it's unfortunately O(n^2) in memory.

I'm also curious to test your proposed fix for the existing algorithm from [this comment](https://github.com/networkx/networkx/issues/4942#issuecomment-873127745). You mentioned changing line 276 to `my_ancestors = nx.dag.ancestors(dag, v)`. I think this may refer to line 272 https://github.com/networkx/networkx/blob/bcc7a7e42b6e152ac57658e884f5184b33463b0b/networkx/algorithms/lowest_common_ancestors.py#L272 in the current version, but I couldn't figure out what needed to be changed?

**Comment by dschult** [comment]
Nice speedups!  :)  Do you understand why the number of LCAs is reported as different for the current and naive algorithm?

You are correct that the line number is now `272` and the suggestion is mistakenly written as exactly the same as the current code !oops!.   It should replace `dag` with `G`:
```python
my_ancestors = nx.dag.ancestors(G, v)
```  
Note that I don't have a proof that this is the correct implementation -- just another way to try to make sense of the Bender paper's description.  The Naive approach is looking better -- more readable and also performant. We could check it's scaling by repeatedly doubling the graph size I suppose.

**Comment by jamestrimble** [comment]
I'm afraid I haven't got my head round the Bender algorithm yet, but the good news is that when I changed line 272 to the version you suggest, it always agrees with the naive algorithm.

The naive approach seems to scale fine - staying about an order of magnitude faster than Bender - although I haven't tried it on anything that takes more than about 10 seconds to solve so I should probably do a larger experiment!

**Comment by jamestrimble** [comment]
Sorry to be putting so many code listings in this thread! Below is the latest version of my naive algorithm implementation, for the record. I think there might even be a citation on Wikipedia to a paper that describes this kind of algorithm.

I'd like to do a proper pull request with a tidied-up version of this algorithm (perhaps as an additional option for a lowest common ancestor algorithm), but I don't have time at the moment. So it would be fantastic if someone else would like to do the PR.

```python
import random
import time
import networkx as nx

def find_successor(G, v, S):
    """Find a successor of v that is in the set of nodes S"""
    for w in G.successors(v):
        if w in S:
            return w
    return None

def get_a_lowest_common_ancestor(G, common_ancestors):
    """Return a lowest common ancestor from the given set of common ancestors"""
    # Start with an arbitrary node v from the set of common ancestors.  Follow
    # arbitrary outgoing edges, remaining in the set of common ancestors, until
    # reaching a node with no outgoing edge to another of the common ancestors.
    v = next(iter(common_ancestors))
    while True:
        successor = find_successor(G, v, common_ancestors)
        if successor is None:
            return v
        v = successor

def get_ancestor_set(G, v, ancestor_cache):
    """Return a set containing v and all of its ancestors in G"""
    if ancestor_cache is not None:
        if v in ancestor_cache:
            return ancestor_cache[v]
    ancestor_set = set([v])
    to_visit = [v]
    while to_visit:
        w = to_visit.pop()
        for u in G.predecessors(w):
            if u not in ancestor_set:
                ancestor_set.add(u)
                to_visit.append(u)
    if ancestor_cache is not None:
        ancestor_cache[v] = ancestor_set
    return ancestor_set

def all_pairs_lowest_common_ancestor(G, pairs=None, use_cache=False):
    """Return the lowest common ancestor of all pairs or the provided pairs"""
    result = []
    ancestor_cache = {} if use_cache else None
    if pairs is None:
        for i, v in enumerate(G.nodes()):
            v_ancestor_set = get_ancestor_set(G, v, ancestor_cache)
            for j, w in enumerate(G.nodes()):
                if j > i:
                    break
                w_ancestor_set = get_ancestor_set(G, w, ancestor_cache)
                common_ancestors = v_ancestor_set & w_ancestor_set
                if common_ancestors:
                    result.append(((v, w), get_a_lowest_common_ancestor(G, common_ancestors)))
    else:
        for v, w in pairs:
            v_ancestor_set = get_ancestor_set(G, v, ancestor_cache)
            w_ancestor_set = get_ancestor_set(G, w, ancestor_cache)
            common_ancestors = v_ancestor_set & w_ancestor_set
            if common_ancestors:
                result.append(((v, w), get_a_lowest_common_ancestor(G, common_ancestors)))
    return result
```

**Comment by dschult** [comment]
Check out #5736 which could be a function that solves this issue once we rename it as the primary function.
Basically, we are giving up on the Bender paper and going to a more obvious (and actually faster) method.
